design/01_proposal.md

```diff
@@ -0,0 +1,71 @@
+# Cli backup proposal
+
+## Context
+
+Personal backups are complex and time-consuming. The goal of this project is to create a simple and easy-to-use tool to backup files and directories. This tool does not aim to just perform backups, but also to guide the user in the process of determining what to backup and how.
+
+## Business concepts
+
+### General concepts
+
+As a user, I have several **devices** (my linux laptop, my server, my S3 bucket, my USB keys, my hard drives, my github account...). I may have valuable data on each of these devices. The part of the data of each device that is the reference one in the **primary data**. For instance, my SSH keys are primarily on my laptop, but my photos are primarily on my hard drive, and my app datas are on my server.
```

Comment:
Maybe clarify a little bit the principle behind what you say : every bit of data is linked to its primary device (and is thus the device's primary data ?), and can be replicated on other devices, the other devices becoming the secondary device for the data (and the device being the secondary device for the data ?)

Reply:
I think it's more of a usage concept here. Even if the data is replicated, the primary is the one where the data is directly accessible and used. The secondary is the device where the data is replicated, but not used. I try another wording.

Reply:
Agreed, but we could thus precise a little bit what it means to be replicated but not used. 
For example if I have a file on my S3, its primary device would be my S3. I can duplicate it on my laptop, my laptop becoming a secondary device for my data. 
So what is the behavior if I want to access my file from my computer ? It should fetch it directly from the S3 ? Or take the one that is on a secondary device (my laptop in this case) ? 

At this point it is really a conception question, but I think it is good to keep that in mind

Reply:
I think it would be much easier to make the secondary totally unusable, so we avoid the risk of corrupting it by user's actions. Compression will for instance bloc direct access to the files on the secondary. What do you think ?

Reply:
I agree for the protection of the data ! 

But with that I think we should handle carefully the place of the personal laptop and/or the device from which we use the CLI ! How should the tool handle if we try to open a file from our S3 from our laptop ? Create a temporary copy of the file and open this copy ? Replace the copy on our laptop (zipped), decompress it and open it ?

Reply:
So, let's assume I have a S3, with some big files "model_xyz.bin". I consider them the primary, because it is the place used by my applications to load the models. 

No, if I want to backup them on my laptop, they might be downloaded from the S3 to a location `.hibernacli/secondaries/s3.../models.zip` . 

If I need to use regularly the models on my laptop, it would have nothing to do with the backup. We would access directly s3 to download the wanted files. 

But if my s3 is deleted after a wrong manipulation, I can restore it using the CLI to reupload models.zip to the files in s3. 

This way I don't think there is scenario where the users mix the secondary with their own workflows.